### ISSUE TITLE
[QUESTION]Megatron-FSDP ZeRO-2 (optim_grads) has higher peak memory than ZeRO-1/ZeRO-3 due to unsharded temporary main_grad bucket allocation

### ISSUE THREAD

**Issue by RayXL97** [NVIDIA/Megatron-LM#6314]
**Describe the question**
When using Megatron-FSDP with --data-parallel-sharding-strategy optim_grads (ZeRO-2) and --gradient-accumulation-fusion (enabled by default), peak memory usage is significantly higher than ZeRO-1 and ZeRO-3. This can cause OOM under ZeRO-2 even when the same configuration runs successfully under ZeRO-1 or ZeRO-3.

Root cause: main_grad_buffer is sharded under ZeRO-2 (is_data_distributed=True), but fetch_bucket() needs to return an unsharded full bucket. Since the pre-allocated buffer is only 1/dp_size in size, the code falls into the else branch and calls allocate_bucket_storage(shard=False), which allocates a temporary buffer sized to self.bucket_index.size — the entire parameter group — rather than per-parameter size. When all parameters land in a single FSDP group, this temporary allocation is roughly the full model size in bf16, and it occurs during backward while activations are still live, driving up peak memory.

**Comparison:**

ZeRO-1 (optim): main_grad_buffer.is_data_distributed=False, fetch_bucket() returns the pre-allocated full buffer directly with zero extra allocation. Peak memory is well-controlled.
ZeRO-3 (optim_grads_params): The same temporary allocation exists, but model weights are also sharded, so baseline memory is lower. Peak memory stays within acceptable range.
ZeRO-2 (optim_grads): Model weights are NOT sharded (full), but main_grad_buffer IS sharded. It has neither ZeRO-1's pre-allocated full buffer nor ZeRO-3's weight sharding savings. During backward: activations + full temporary bucket + full model weights all coexist, giving the highest peak memory of the three strategies.

**Root cause code path**

tensor_parallel/layers.py — backward calls weight.main_grad = weight.get_main_grad()
param_and_grad_buffer.py — main_grad_getter calls p._gbuf.fetch_bucket(dtype=grad_comm_dtype) (where p._gbuf = main_grad_buffer)
param_and_grad_buffer.py — Under ZeRO-2, is_data_distributed=True, so not self.is_data_distributed is False, falling into else: self.allocate_bucket_storage()
param_and_grad_buffer.py — allocate_bucket_storage(shard=False) uses self.bucket_index.size (full group size), not the sharded size
param_and_grad_buffer.py — torch.empty(full_bucket_size, dtype=bf16) allocates a large temporary buffer
Relevant flag setting in param_and_grad_buffer.py:

# ZeRO-1:
is_grad_buffer_distributed = False  # main_grad_buffer is full-size, pre-allocated

# ZeRO-2:
is_grad_buffer_distributed = True   # main_grad_buffer is sharded → fetch_bucket()
                                     #   falls into else → allocates full temp bucket
Why gradient_accumulation_fusion matters

With fusion (default): get_main_grad() is called during backward while activations are still live. The temporary bucket coexists with activation memory.
Without fusion: get_main_grad() is deferred until after backward completes, when activations have already been freed. Peak memory is lower.

**Comment by janEbert** [comment]
@cspades could you take a look?

> When all parameters land in a single FSDP group, this temporary allocation is roughly the full model size in bf16, and it occurs during backward while activations are still live, driving up peak memory.

This definitely doesn't seem desired.

**Comment by cspades** [comment]
The claim is that ZeRO-1 uses less memory than ZeRO-2, everything else mentioned is not a bug, we must allocate unsharded gradient buffers temporarily for reduce-scatter. ZeRO-1 does not need to do this because the gradient is unsharded anyway, so we just use the unsharded gradient buffer as the reduce-scatter buffer.

> When all parameters land in a single FSDP group

You need to specify all the unit modules that you want to shard/unshard temporarily. Otherwise, the whole model is an FSDP unit and will be simultaneously unsharded, in which case Z1 will use less memory than Z2. If your model is a single layer, then FSDP is not the answer, use Tensor Parallelism.

@RayXL97 please provide evidence and reproduction scripts so we can reproduce the claim, thanks!

**Comment by RayXL97** [comment]
use two A100(40G) to test ,if SHARDING_STRATEGY is "optim", it will work;if SHARDING_STRATEGY is "optim_grads", it will OOM 
# Distributed training variables
NNODES=1
GPUS_PER_NODE=2
GPU_NUM=$((${GPUS_PER_NODE}*${NNODES}))
WORLD_SIZE=$((${GPUS_PER_NODE}*${NNODES}))
NODE_RANK=0
MASTER_PORT=12453
MASTER_ADDR=localhost

# Parallelism variables
TP=1
USP=1
PP=1
DP=$((${GPU_NUM}/${TP}/${PP}))

# Network size variables
MODEL_SIZE=7

if   [ ${MODEL_SIZE} == 7 ];   then HIDDEN_SIZE=4096;  NUM_HEAD=32; NUM_QUERY_GROUP=32; NUM_LAYERS=10; FFN_HIDDEN_SIZE=11008; NORM_EPS=1e-5;
elif [ ${MODEL_SIZE} == 13 ];  then HIDDEN_SIZE=5120;  NUM_HEAD=40; NUM_QUERY_GROUP=40; NUM_LAYERS=40; FFN_HIDDEN_SIZE=13824; NORM_EPS=1e-5;
elif [ ${MODEL_SIZE} == 70 ];  then HIDDEN_SIZE=8192;  NUM_HEAD=64; NUM_QUERY_GROUP=8;  NUM_LAYERS=2; FFN_HIDDEN_SIZE=28672; NORM_EPS=1e-5;
elif [ ${MODEL_SIZE} == 130 ];  then HIDDEN_SIZE=12288;  NUM_HEAD=96; NUM_QUERY_GROUP=8;  NUM_LAYERS=88; FFN_HIDDEN_SIZE=31232; NORM_EPS=1e-5;
elif [ ${MODEL_SIZE} == "tiny" ]; then HIDDEN_SIZE=128;  NUM_HEAD=4; NUM_QUERY_GROUP=4; NUM_LAYERS=4; FFN_HIDDEN_SIZE=512; NORM_EPS=1e-5;
else echo "invalid MODEL_SIZE: ${MODEL_SIZE}"; exit 1
fi

DROP_OUT=0.0
MAX_SEQ_LEN=4096
MAX_POSITION_EMBEDDINGS=4096

SRC_PATH=../pretrain_gpt.py

DATA_PATH=/external/pcie-softwar/data/oscar-en-10k-meg-llama_text_document
MODEL_PATH=/external/pcie-software/model/tokenizer.model


# Set training command
LAUNCHER=" \
       torchrun \
       --nproc_per_node ${GPUS_PER_NODE} \
       --nnodes ${NNODES} \
       --node_rank ${NODE_RANK} \
       --master_addr ${MASTER_ADDR} \
       --master_port ${MASTER_PORT} \
       "


DISTRIBUTED_ARGS=" \
       --tensor-model-parallel-size ${TP} \
       --pipeline-model-parallel-size ${PP} \
       --distributed-backend nccl \
       --use-distributed-optimizer \
       --sequence-parallel \
       "    

NETWORK_SIZE_ARGS=" \
       --num-layers ${NUM_LAYERS} \
       --hidden-size ${HIDDEN_SIZE} \
       --num-attention-heads ${NUM_HEAD} \
       --ffn-hidden-size ${FFN_HIDDEN_SIZE} \
       --max-position-embeddings ${MAX_POSITION_EMBEDDINGS} \
       --norm-epsilon ${NORM_EPS} \
       --untie-embeddings-and-output-weights \
       --no-masked-softmax-fusion \
       --sequence-parallel \
       --group-query-attention \
       --num-query-groups ${NUM_QUERY_GROUP} \
       --swiglu \
       --normalization RMSNorm \
       --use-flash-attn \
       --attention-backend flash \
       "

LOGGING_ARGS=" \
       --log-timers-to-tensorboard \
       --log-validation-ppl-to-tensorboard \
       --log-memory-to-tensorboard \
       "

REGULATIZATION_ARGS=" \
       --attention-dropout ${DROP_OUT} \
       --hidden-dropout ${DROP_OUT} \
       --weight-decay 1e-1 \
       --clip-grad 1.0 \
       --adam-beta1 0.9 \
       --adam-beta2 0.95 \
       --adam-eps 1e-8 \
       "

SHARDING_STRATEGY=${SHARDING_STRATEGY:-"optim"}
TRAINING_ARGS=" \
       --micro-batch-size 1 \
       --global-batch-size 64 \
       --train-iters 20 \
       --log-interval 1 \
       --optimizer adam \
       --use-megatron-fsdp \
       --data-parallel-sharding-strategy ${SHARDING_STRATEGY}
       "

RECOMPUTE_ARGS="
"


INITIALIZATION_ARGS=" \
       --seed 1024 \
       --init-method-std 0.02 \
       "

LEARNING_RATE_ARGS=" \
       --lr 3e-4 \
       --lr-decay-style cosine \
       --lr-warmup-fraction 0.01 \
       --min-lr 3e-5 \
       "

CHECKPOINTING_ARGS=" \
       --finetune \
       --no-load-optim \
       --no-load-rng \
       --save-interval 1000000 \
       "

MIXED_PRECISION_ARGS=" \
       --bf16 \
       --initial-loss-scale 65536.0 \
       --min-loss-scale 1.0 \
       --loss-scale-window 1000 \
       "

VALIDATION_ARGS=" \
       --eval-interval 10000 \
       --eval-iters 10 \
       "

DATA_ARGS=" \
       --data-path ${DATA_PATH} \
       --split 949,50,1 \
       --seq-length ${MAX_SEQ_LEN} \
       --num-workers 4 \
       --tokenizer-type Llama2Tokenizer \
       --tokenizer-model ${MODEL_PATH} \
       --dataloader-type single \
       --ckpt-format fsdp_dtensor \
       "

CMD="${LAUNCHER} \
       ${SRC_PATH} \
       ${DISTRIBUTED_ARGS} \
       ${NETWORK_SIZE_ARGS} \
       ${LOGGING_ARGS} \
       ${REGULATIZATION_ARGS} \
       ${TRAINING_ARGS} \
       ${RECOMPUTE_ARGS} \
       ${INITIALIZATION_ARGS} \
       ${LEARNING_RATE_ARGS} \
       ${CHECKPOINTING_ARGS} \
       ${MIXED_PRECISION_ARGS} \
       ${VALIDATION_ARGS} \
       ${DATA_ARGS} \
       ${MOE_ARGS} \
       "
echo ${CMD}
${CMD}

**Comment by cspades** [comment]
@RayXL97 This is pretty much a bug that I think you can fix by hard-coding this change: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/mcore_fsdp_adapter.py#L172
```
        if fsdp_unit_modules is not None:
            self.fsdp_unit_modules = fsdp_unit_modules
        else:
            if self.ddp_config.data_parallel_sharding_strategy != "no_shard":
                self.fsdp_unit_modules = [TransformerLayer, MoETransformerLayer, MambaLayer]
            else:
                self.fsdp_unit_modules = []
```

Don't ask me why people have been hard-coding this. And all my ZeRO-2 tests are in `fully_shard` not the MLM path that hard-codes it, I guess the assumption is that nobody uses ZeRO-2 in Megatron-LM. Perhaps it is missable because even if you use the broken ZeRO-2 without the right sharding units, the max allocated and total memory is still less than ZeRO-1 in many cases (mostly due to a cross-entropy activation spike that hides the LM embedding gradient allocation, which is why the peak memory is only like 300 MB less even after the fix):

```
# ZeRO-1
[Rank 0] (after 5 iterations) memory (MB) | allocated: 25075.44 | max allocated: 32346.11 | reserved: 27338.00 | max reserved: 34526.00 | total device memory used: 32639.25

# ZeRO-2 (Before Fix)
[Rank 0] (after 5 iterations) memory (MB) | allocated: 20651.87 | max allocated: 29844.42 | reserved: 23172.00 | max reserved: 39896.00 | total device memory used: 28473.25

# ZeRO-2 (After Fix)
[Rank 0] (after 5 iterations) memory (MB) | allocated: 20693.61 | max allocated: 27960.13 | reserved: 22882.00 | max reserved: 32142.00 | total device memory used: 28183.25
```

What you've been seeing:

<img width="2292" height="1023" alt="Image" src="https://github.com/user-attachments/assets/b360bc1a-0d1e-4d63-9494-112467abfd1c" />

and what it should be:

<img width="2160" height="1053" alt="Image" src="https://github.com/user-attachments/assets/b3bd5efa-12e7-4cd4-a643-2e4a77523c74" />

**Comment by cspades** [comment]
Also practically speaking, I don't think it matters too much whether we unshard the full model or just a unit at a time (the floor is the same, all we save is how much memory we unshard at a time) because of that cross-entropy spike. You're likely better just going ZeRO-3 and sharding as much as you can to lower the memory floor.

<img width="1179" height="1065" alt="Image" src="https://github.com/user-attachments/assets/e5a5a1ed-d1e4-4a90-a389-ceb32050f45e" />

**Comment by cspades** [comment]
@RayXL97 Thanks for raising this, PR is up: https://github.com/NVIDIA/Megatron-LM/pull/6357

I ended up finding another bug (completely screwed up gradients) with Z1 + wgrad accumulate fusion (which you were using, turned off with `--no-gradient-accumulation-fusion`) that I also fixed:
```
# Optim w/ Gradient Accum Fusion - main / BUGGED
[2026-08-07 12:12:18.986572] iteration       70/  305175 | consumed samples:         4480 | elapsed time per iteration (ms): 1109.2 | throughput per GPU (TFLOP/s/GPU): 963.1 | learning rate: 1.720342E-05 | global batch size:    64 | lm loss: 6.889314E+00 | loss scale: 1.0 | grad norm: 0.881 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |

# Optim (Overlapped Reduce-Scatter) w/ Gradient Accum Fusion - Fixed with known slight discrepancy when accumulating into a pre-allocated gradient buffer.
[2026-08-07 14:19:47.622783] iteration       70/  305175 | consumed samples:         4480 | elapsed time per iteration (ms): 1098.5 | throughput per GPU (TFLOP/s/GPU): 972.6 | learning rate: 1.720342E-05 | global batch size:    64 | lm loss: 1.242577E+00 | loss scale: 1.0 | grad norm: 5.195 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |

# Optim (Non-Overlapped Reduce-Scatter) w/ Gradient Accum Fusion - Fixed with known slight discrepancy when accumulating into a pre-allocated gradient buffer.
[2026-08-07 14:34:02.532789] iteration       70/  305175 | consumed samples:         4480 | elapsed time per iteration (ms): 1101.3 | throughput per GPU (TFLOP/s/GPU): 970.0 | learning rate: 1.720342E-05 | global batch size:    64 | lm loss: 1.241726E+00 | loss scale: 1.0 | grad norm: 5.300 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |

# Optim w/o Gradient Accum Fusion
[2026-08-07 14:09:07.595266] iteration       70/  305175 | consumed samples:         4480 | elapsed time per iteration (ms): 1138.7 | throughput per GPU (TFLOP/s/GPU): 938.2 | learning rate: 1.720342E-05 | global batch size:    64 | lm loss: 1.168228E+00 | loss scale: 1.0 | grad norm: 5.033 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |

# ZeRO-2
[2026-08-07 12:30:55.867010] iteration       70/  305175 | consumed samples:         4480 | elapsed time per iteration (ms): 1164.8 | throughput per GPU (TFLOP/s/GPU): 917.1 | learning rate: 1.720342E-05 | global batch size:    64 | lm loss: 1.168045E+00 | loss scale: 1.0 | grad norm: 5.035 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |

# FSDP
[2026-08-07 12:25:21.417691] iteration       70/  305175 | consumed samples:         4480 | elapsed time per iteration (ms): 1348.9 | throughput per GPU (TFLOP/s/GPU): 792.0 | learning rate: 1.720342E-05 | global batch size:    64 | lm loss: 1.168174E+00 | loss scale: 1.0 | grad norm: 5.051 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
```

**Comment by RayXL97** [comment]
Thanks!
